README.md
# PgAuth

Auth library depends on Postgresql.

### Support:
- Normal auth and OAuth
- User management
- Role management
- Department management
- App management

### Auth Method
config auth_method
- Login by name and password: 1
- Login by name and access_token: 2
- Login by access_token only: 4
- Login with user_id only: 8
- Login by email and password: 16
- Login by email and access_token: 32
- Login by tel and password: 64
- Login by tel and access_token: 128

And login TTL, config token_ttl
- Login once (config token_ttl with 0, no cache)
- Login for some time (config token_ttl with > 0, use cache)

### Usage
#### Initialize database
`Run \i install.sql in psql`

#### Auth
- Normal Auth
```
$auth = new \Wwtg99\PgAuth\Auth\NormalAuth($conn, $config);
//sign up
$user1 = ['name'=>'u1', 'password'=>'1', 'label'=>'user 1', 'email'=>'u1@a.com'];
$u = $auth->signUp($user1);
var_dump($u->getUser());
//sign in
$user2 = ['username'=>'u1', 'password'=>'1'];
$u = $auth->signIn($user2);
var_dump($u->getUser());
//verify but not sign in
$re = $auth->verify($user2);
var_dump($re);
//sign out
$auth->signOut($u->getUser());

//change user info by user object
$user3 = ['label'=>'user 111', 'descr'=>'aaaa'];
$re = $u->changeInfo($user3);
//inactive user
$re = $u->inactive();
//active user
$re = $u->active();
```

- OAuth
```
$pool = new \Wwtg99\DataPool\Common\DefaultDataPool('../conf.json');
$conn = $pool->getConnection('main');
$config = [
    "cache"=>[
        "type"=>"redis",
        "options"=>[
            "schema"=>"tcp",
            "host"=>"192.168.83.128",
            "database"=>4
        ]
    ],
    'auth_method'=>7,
    'token_ttl'=>60,
    'code_ttl'=>60
];
$appu = new \Wwtg99\PgAuth\Utils\AppUtils($conn);
//step 1 get code
$user1 = ['username'=>'u1', 'password'=>'2'];
$aid = 'asfafs';
$redirect_uri = 'localhost';
$auth = new \Wwtg99\PgAuth\Auth\NormalAuth($conn, $config);
$oauth = new \Wwtg99\PgAuth\Utils\OAuthUtils($config);
$code = '';
$app = $appu->verifyAppIdUri($aid, $redirect_uri);
if ($app) {
    $u = $auth->signIn($user1);
    if ($u) {
        $code = $oauth->generateCode($u->getUserArray(), $aid, $redirect_uri);
        echo "\nCode: $code";
    }
}
//step 2 get access_token
$secret = 'fasf3wfsdf';
$u = $oauth->verifyCode($code);
if ($u) {
    $aid = $u['app_id'];
    $re = $appu->verifySecret($aid, $secret);
    if ($re) {
        $token = $u['access_token'];
        echo "\nUser: ";
        print_r($u);
    }
}
```

- Manage department (One user belongs to one department or not)
```
// create department
$dep = $this->pool->getConnection('main')->getMapper('Department');
$data = ['department_id'=>'d1', 'name'=>'dep1', 'descr'=>'dddd'];
$re = $dep->insert($data);
// update department
$data_update = ['name'=>'dep11', 'descr'=>'ffff'];
$re = $dep->update($data_update, null, 'd1');
// get all departments
$re = $dep->select();
// get department by id
$re = $dep->get('d1');
// delete department
$re = $dep->delete('d1');
```
- Manage role (One user can belong to many roles)
```
// create role
$role = $this->pool->getConnection('main')->getMapper('Role');
$data = ['name'=>'r1', 'descr'=>'rrr'];
$rid = $role->insert($data);
// update role
$data_update = ['name'=>'role1', 'descr'=>'ggg'];
$re = $role->update($data_update, null, $rid);
// get all roles
$re = $role->select();
// get role by id
$re = $role->get($rid);
// delete role
$re = $role->delete($rid);
```
- Manage user
```
// create user
$user = $this->pool->getConnection('main')->getMapper('User');
$data = ['name'=>'u1', 'label'=>'user1', 'email'=>'u@u.com', 'descr'=>'rrr', 'department_id'=>'d1'];
$uid = $user->insert($data);
// update user
$data_update = ['name'=>'user1', 'descr'=>'ggg'];
$re = $user->update($data_update, null, $uid);
// get all user
$re = $user->select();
// get user by id
$re = $user->get($uid);
// add role
$re = $user->addRole($uid, 1);
// remove role
$re = $user->removeRole($uid, 1);
// change roles
$re = $user->changeRoles($uid, [['role_name'=>'role1']]);
// user view (get all info of user)
$re = $user->view(); //view all
$re = $user->view('*', ['user_id'=>$uid]); //view by user_id
// delete user
$re = $user->delete($uid);
```
- Manage app
```
// create app
$app = $this->pool->getConnection('main')->getMapper('App');
$data = ['app_name'=>'app1', 'descr'=>'rrr', 'redirect_uri'=>'localhost'];
$aid = $app->insert($data);
// update app
$data_update = ['app_name'=>'app11', 'descr'=>'ggg'];
$re = $app->update($data_update, null, $aid);
// get all app
$re = $app->select();
// get app by id
$re = $app->get($aid);
// delete app
$re = $app->delete($aid);
```


CHANGELOG.md
## 0.2.1
- Update view_users
- Update NormalAuth verify password methods

## 0.2.0
- Change OAuth to utils
- Add more verify methods
- Update class structures

## 0.1.10
- Fix auth_method bugs

## 0.1.9
- Add auth_method
- Support login by user_id
- Remove client

## 0.1.8
- Add getUser() in IAuth

## 0.1.7
- Simplify auth and user
- Update db structure

## 0.1.6
- Fix refresh user after update info

## 0.1.5
- Fix NormalUser bugs
- Update sql

## 0.1.4
- Remove roles add common_user
- Update pgauth-cli
- Update OAuthServer

## 0.1.3
- Update OAuthServer keys to be configurable

## 0.1.2
- Add json params to departments, roles and apps
- Add Auth and User
- Support NormalAuth and NormalUser
- Add pgauth-cli

## 0.1.1
- Add params to users

## 0.1.0
- Add App, Department, Role, User, UserLog

src/Auth/NormalAuth.php
<?php
/**
 * Created by PhpStorm.
 * User: wwt
 * Date: 2016/10/6 0006
 * Time: 下午 4:02
 */

namespace Wwtg99\PgAuth\Auth;


use Wwtg99\DataPool\Common\IDataConnection;
use Wwtg99\PgAuth\Mapper\User;
use Wwtg99\PgAuth\Utils\Cache;

/**
 * Class NormalAuth
 * Normal auth
 * Verify by username, user_id, password or access_token.
 * @package Wwtg99\PgAuth\Auth
 */
class NormalAuth extends AbstractAuth
{

    const VERIFY_NAME_PASSWORD = 1;
    const VERIFY_NAME_TOKEN = 2;
    const VERIFY_TOKEN = 4;
    const VERIFY_USER_ID = 8;
    const VERIFY_EMAIL_PASSWORD = 16;
    const VERIFY_EMAIL_TOKEN = 32;
    const VERIFY_TEL_PASSWORD = 64;
    const VERIFY_TEL_TOKEN = 128;

    /**
     * @var \Desarrolla2\Cache\Cache
     */
    protected $cache = null;

    /**
     * @var \Wwtg99\DataPool\Common\IDataConnection
     */
    protected $conn = null;

    /**
     * Verify method
     * 1: name and password
     * 2: name and access_token
     * 4: access_token only
     * 8: user_id only
     * 16: email and password
     * 32: email and access_token
     * 64: tel and password
     * 128: tel and access_token
     * @var int
     */
    protected $authMethod = 1;

    /**
     * @var int
     */
    protected $tokenTtl = 3600;

    /**
     * @var bool
     */
    protected $uniqueEmail = true;

    /**
     * @var bool
     */
    protected $uniqueTel = true;

    /**
     * NormalAuth constructor.
     *
     * @param IDataConnection $conn
     * @param array $config
     */
    public function __construct($conn = null, $config = null)
    {
        if (isset($config['cache'])) {
            $ch = $config['cache'];
            if (isset($ch['type']) && isset($ch['options'])) {
                $this->cache = Cache::getCache($ch['type'], $ch['options']);
            }
        }
        if (isset($config['token_ttl'])) {
            $this->tokenTtl = $config['token_ttl'];
        }
        if (isset($config['auth_method'])) {
            $this->authMethod = $config['auth_method'];
        }
        if (isset($config['unique_email'])) {
            $this->uniqueEmail = boolval($config['unique_email']);
        }
        if (isset($config['unique_tel'])) {
            $this->uniqueTel = boolval($config['unique_tel']);
        }
        $this->conn = $conn;
    }

    /**
     * @param array $user
     * @return IUser|null
     */
    public function signUp(array $user)
    {
        $userModel = $this->conn->getMapper('User');
        //check exists
        $wh = [IUser::FIELD_USER_NAME=>$user[IUser::FIELD_USER_NAME]];
        if ($this->uniqueEmail && isset($user[IUser::FIELD_EMAIL]) && $user[IUser::FIELD_EMAIL]) {
            $wh[IUser::FIELD_EMAIL] = $user[IUser::FIELD_EMAIL];
        }
        if ($this->uniqueTel && isset($user[IUser::FIELD_TEL]) && $user[IUser::FIELD_TEL]) {
            $wh[IUser::FIELD_TEL] = $user[IUser::FIELD_TEL];
        }
        $re = $userModel->has(['OR'=>$wh]);
        if ($re) {
            $this->msg = 'User ' . $user[IUser::FIELD_USER_NAME] . ' exists!';
            return null;
        }
        //check roles
        if (isset($user[IUser::FIELD_ROLES])) {
            $roles = $user[IUser::FIELD_ROLES];
            if (!is_array($roles)) {
                $roles = explode(',', $roles);
            }
            $r = [];
            foreach ($roles as $role) {
                array_push($r, ['role_name'=>$role]);
            }
            unset($user[IUser::FIELD_ROLES]);
        }
        //check password
        $pwd = isset($user[IUser::FIELD_PASSWORD]) ? $user[IUser::FIELD_PASSWORD] : null;
        if ($pwd) {
            $pwd = password_hash($pwd, PASSWORD_BCRYPT);
        }
        $user[IUser::FIELD_PASSWORD] = $pwd;
        //insert user
        $uid = $userModel->insert($user);
        if ($uid) {
            if (isset($r)) {
                $re = $userModel->changeRoles($uid, $r);
                if (!$re) {
                    $this->msg = 'Change roles failed!';
                }
            }
            $token = $this->generateToken($uid);
            $this->user = new NormalUser($uid, $userModel, $token);
            $this->msg = 'Sign up successfully!';
            if ($this->tokenTtl > 0) {
                $this->saveCache();
            }
            return $this->user;
        }
        $this->msg = 'Sign up failed!';
        return null;
    }

    /**
     * @param array $user
     * @return IUser|null
     */
    public function signIn(array $user)
    {
        if ($this->verify($user)) {
            $this->msg = 'Sign in successfully!';
            if ($this->tokenTtl > 0) {
                $this->saveCache();
            }
        }
        return $this->user;
    }

    /**
     * @param array $user
     * @return IUser|null
     */
    public function signOut(array $user)
    {
        $token = isset($user[self::KEY_TOKEN]) ? $user[self::KEY_TOKEN] : null;
        if ($token) {
            $this->cache->delete($token);
        }
        $this->msg = 'Sign out successfully!';
        return null;
    }

    /**
     * @param array $user
     * @return bool
     */
    public function verify(array $user)
    {
        $re = false;
        if (($this->authMethod & 4) == 4 && isset($user[self::KEY_TOKEN])) {
            //check access token only
            $re = $this->checkTokenOnly($user);
        }
        if (!$re && ($this->authMethod & 2) == 2 && isset($user[self::KEY_TOKEN]) && isset($user[self::KEY_USERNAME])) {
            //check access token and name
            $re = $this->checkTokenName($user);
        }
        if (!$re && ($this->authMethod & 32) == 32 && isset($user[self::KEY_TOKEN]) && isset($user[self::KEY_EMAIL])) {
            //check access token and email
            $re = $this->checkTokenEmail($user);
        }
        if (!$re && ($this->authMethod & 128) == 128 && isset($user[self::KEY_TOKEN]) && isset($user[self::KEY_TEL])) {
            //check access token and tel
            $re = $this->checkTokenTel($user);
        }
        if (!$re && ($this->authMethod & 1) == 1 && isset($user[self::KEY_USERNAME]) && isset($user[self::KEY_PASSWORD])) {
            //check name and password
            $re = $this->checkNamePassword($user);
        }
        if (!$re && ($this->authMethod & 16) == 16 && isset($user[self::KEY_EMAIL]) && isset($user[self::KEY_PASSWORD])) {
            //check email and password
            $re = $this->checkEmailPassword($user);
        }
        if (!$re && ($this->authMethod & 64) == 64 && isset($user[self::KEY_TEL]) && isset($user[self::KEY_PASSWORD])) {
            //check tel and password
            $re = $this->checkTelPassword($user);
        }
        if (!$re && ($this->authMethod & 8) == 8 && isset($user[self::KEY_USER_ID])) {
            //check user_id
            $re = $this->checkUserId($user);
        }
        return $re;
    }

    /**
     * @return \Desarrolla2\Cache\Cache
     */
    public function getCache()
    {
        return $this->cache;
    }

    /**
     * Save token in cache
     */
    public function saveCache()
    {
        if ($this->user) {
            $token = $this->user->getUserArray()[IUser::FIELD_TOKEN];
            //store token in cache
            $this->cache->set($token, json_encode($this->user->getUserArray(), JSON_UNESCAPED_UNICODE), $this->tokenTtl);
        }
    }

    /**
     * @return int
     */
    public function getTokenTtl()
    {
        return $this->tokenTtl;
    }

    /**
     * @param int $tokenTtl
     * @return NormalAuth
     */
    public function setTokenTtl($tokenTtl)
    {
        $this->tokenTtl = $tokenTtl;
        return $this;
    }

    /**
     * @return boolean
     */
    public function isUniqueEmail()
    {
        return $this->uniqueEmail;
    }

    /**
     * @param boolean $uniqueEmail
     * @return NormalAuth
     */
    public function setUniqueEmail($uniqueEmail)
    {
        $this->uniqueEmail = $uniqueEmail;
        return $this;
    }

    /**
     * @return boolean
     */
    public function isUniqueTel()
    {
        return $this->uniqueTel;
    }

    /**
     * @param boolean $uniqueTel
     * @return NormalAuth
     */
    public function setUniqueTel($uniqueTel)
    {
        $this->uniqueTel = $uniqueTel;
        return $this;
    }

    /**
     * @return int
     */
    public function getAuthMethod()
    {
        return $this->authMethod;
    }

    /**
     * @param int $authMethod
     * @return NormalAuth
     */
    public function setAuthMethod($authMethod)
    {
        $this->authMethod = $authMethod;
        return $this;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkTokenOnly($user)
    {
        $token = $user[self::KEY_TOKEN];
        if ($this->cache->has($token)) {
            $u = $this->cache->get($token);
            if ($u) {
                $uobj = json_decode($u, true);
                if (isset($uobj[IUser::FIELD_USER_ID])) {
                    $uid = $uobj[IUser::FIELD_USER_ID];
                    $userModel = $this->conn->getMapper('User');
                    $this->user = new NormalUser($uid, $uobj, $token);
                    if ($userModel instanceof User) {
                        $this->user->setMapper($userModel);
                    }
                    $this->msg = 'User is valid!';
                    return true;
                }
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkTokenName($user)
    {
        $token = $user[self::KEY_TOKEN];
        if ($this->cache->has($token)) {
            $u = $this->cache->get($token);
            if ($u) {
                $uobj = json_decode($u, true);
                $uname = $user[self::KEY_USERNAME];
                if (isset($uobj[IUser::FIELD_USER_ID]) && $uname == $uobj[IUser::FIELD_USER_NAME]) {
                    $uid = $uobj[IUser::FIELD_USER_ID];
                    $userModel = $this->conn->getMapper('User');
                    $this->user = new NormalUser($uid, $uobj, $token);
                    if ($userModel instanceof User) {
                        $this->user->setMapper($userModel);
                    }
                    $this->msg = 'User is valid!';
                    return true;
                }
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkTokenEmail($user)
    {
        $token = $user[self::KEY_TOKEN];
        if ($this->cache->has($token)) {
            $u = $this->cache->get($token);
            if ($u) {
                $uobj = json_decode($u, true);
                $email = $user[self::KEY_EMAIL];
                if (isset($uobj[IUser::FIELD_USER_ID]) && $email == $uobj[IUser::FIELD_EMAIL]) {
                    $uid = $uobj[IUser::FIELD_USER_ID];
                    $userModel = $this->conn->getMapper('User');
                    $this->user = new NormalUser($uid, $uobj, $token);
                    if ($userModel instanceof User) {
                        $this->user->setMapper($userModel);
                    }
                    $this->msg = 'User is valid!';
                    return true;
                }
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkTokenTel($user)
    {
        $token = $user[self::KEY_TOKEN];
        if ($this->cache->has($token)) {
            $u = $this->cache->get($token);
            if ($u) {
                $uobj = json_decode($u, true);
                $tel = $user[self::KEY_TEL];
                if (isset($uobj[IUser::FIELD_USER_ID]) && $tel == $uobj[IUser::FIELD_TEL]) {
                    $uid = $uobj[IUser::FIELD_USER_ID];
                    $userModel = $this->conn->getMapper('User');
                    $this->user = new NormalUser($uid, $uobj, $token);
                    if ($userModel instanceof User) {
                        $this->user->setMapper($userModel);
                    }
                    $this->msg = 'User is valid!';
                    return true;
                }
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkNamePassword($user)
    {
        $userModel = $this->conn->getMapper('User');
        $u = $userModel->view('*', ['AND'=>[IUser::FIELD_USER_NAME => $user[self::KEY_USERNAME], 'deleted_at'=>null]]);
        if ($u && isset($u[0])) {
            $u = $u[0];
            $pwd = $u[IUser::FIELD_PASSWORD];
            if (is_null($pwd) || password_verify($user[self::KEY_PASSWORD], $pwd)) {
                $uid = $u[IUser::FIELD_USER_ID];
                $token = $this->generateToken($uid);
                unset($u[IUser::FIELD_PASSWORD]);
                $this->user = new NormalUser($uid, $u, $token);
                if ($userModel instanceof User) {
                    $this->user->setMapper($userModel);
                }
                $this->msg = 'User is valid!';
                return true;
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkEmailPassword($user)
    {
        $userModel = $this->conn->getMapper('User');
        $u = $userModel->view('*', ['AND'=>[IUser::FIELD_EMAIL => $user[self::KEY_EMAIL], 'deleted_at'=>null]]);
        if ($u && isset($u[0])) {
            $u = $u[0];
            $pwd = $u[IUser::FIELD_PASSWORD];
            if (is_null($pwd) || password_verify($user[self::KEY_PASSWORD], $pwd)) {
                $uid = $u[IUser::FIELD_USER_ID];
                $token = $this->generateToken($uid);
                unset($u[IUser::FIELD_PASSWORD]);
                $this->user = new NormalUser($uid, $u, $token);
                if ($userModel instanceof User) {
                    $this->user->setMapper($userModel);
                }
                $this->msg = 'User is valid!';
                return true;
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkTelPassword($user)
    {
        $userModel = $this->conn->getMapper('User');
        $u = $userModel->view('*', ['AND'=>[IUser::FIELD_TEL => $user[self::KEY_TEL], 'deleted_at'=>null]]);
        if ($u && isset($u[0])) {
            $u = $u[0];
            $pwd = $u[IUser::FIELD_PASSWORD];
            if (is_null($pwd) || password_verify($user[self::KEY_PASSWORD], $pwd)) {
                $uid = $u[IUser::FIELD_USER_ID];
                $token = $this->generateToken($uid);
                unset($u[IUser::FIELD_PASSWORD]);
                $this->user = new NormalUser($uid, $u, $token);
                if ($userModel instanceof User) {
                    $this->user->setMapper($userModel);
                }
                $this->msg = 'User is valid!';
                return true;
            }
        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $user
     * @return bool
     */
    protected function checkUserId($user)
    {
        $userModel = $this->conn->getMapper('User');
        $uid = $user[self::KEY_USER_ID];
        $u = $userModel->view(null, '*', ['AND'=>[IUser::FIELD_USER_ID => $uid, 'deleted_at'=>null]]);
        if ($u && isset($u[0])) {
            $u = $u[0];
            $token = $this->generateToken($uid);
            unset($u[IUser::FIELD_PASSWORD]);
            $this->user = new NormalUser($uid, $u, $token);
            if ($userModel instanceof User) {
                $this->user->setMapper($userModel);
            }
            $this->msg = 'User is valid!';
            return true;

        }
        $this->msg = 'Verify user failed!';
        return false;
    }

    /**
     * @param $uid
     * @return string
     */
    protected function generateToken($uid)
    {
        return substr(md5($uid . mt_rand(0, 1000) . time()), 3, 10);
    }

}